### ISSUE TITLE
How to express implicit dependencies?

### ISSUE THREAD

**Issue by ljharb** [domenic/proposal-import-function#16]
With static imports:

``` js
import 'es5-shim';
import 'es6-shim';
```

theoretically downloads both resources in parallel, but executes them in order - which is good, because es6-shim depends on es5-shim being run first. Said another way, lexical order of static imports also expresses an implicit dependency on execution order.

How can I achieve this with `import()`?

One obvious solution is that I could `import('es5-shim').then(() => import('es6-shim'))` - but then the resources are not downloaded in parallel.

Another obvious solution is that I could create a file, "shims.js", that consisted of the static imports mentioned previously - and then `import('./shims.js')` would issue one request to download that file, followed by two parallel requests that were then executed in order. This isn't bad, but requires a) an additional HTTP request, and b) an additional file.

https://github.com/webpack/webpack/issues/3141 is another example of this same problem using `Intl` polyfills.

Also obviously, implicit dependencies like this are subpar - ideally `es6-shim` should import `es5-shim`, thus making the dependency explicit; and ideally `intl/locale-data/jsonp/en` should import `intl`, thus making _that_ dependency explicit. However, this clearly isn't always possible.

One possibility is wasting future extensibility by making `import()` take multiple arguments, such that arguments would all be downloaded in parallel but executed in order and a single promise for all returned; another is allowing `import()` to take an array as well as a string (which would be a runtime thing, or else it would have to be an array literal) and when an array is passed, the same behavior occurs. However, I'm not sure whether those are a good idea.

Do you think this is something `import()` can solve? If not, what would you suggest as an alternative? If so, what approach do you think would be the least objectionable?

**Comment by domenic** [comment]
> One obvious solution is that I could import('es5-shim').then(() => import('es6-shim')) - but then the resources are not downloaded in parallel.

That's not necessarily true. Engines could "preload scan" for simple forms of import() and start loading in parallel. But yeah, it's not guaranteed.

> Another obvious solution is that I could create a file, "shims.js", that consisted of the static imports mentioned previously

This sounds like the correct solution to me, if you can't just use static import syntax directly. It is most in line with what this spec is designed for---dynamically kicking off the load of a subgraph. Presumably here shims.js is not actually "shims.js": it's "entry-file-for-a-part-of-the-app-that-depends-on-some-shims-to-function.js". So it will have a lot more than those two imports.

> Also obviously, implicit dependencies like this are subpar

Yes, exactly. I think not only are such implicit dependencies outside the scope of this proposal, they're also outside the scope of the entire ES module system.

> One possibility is wasting future extensibility by making import() take multiple arguments

This is indeed an interesting possibility. However, I'm hesitant to explore such things just to support such anti-modular, IMO out of scope, patterns.

> If not, what would you suggest as an alternative?

The best alternative will be using HTTP/2 push or two `<link rel="preload">` / equivalent `Link` headers, to ensure the fetch for those files starts as soon as possible---preferably before the JavaScript file with the `import()` statement is even fetched or parsed itself. Then, when you do the `import('es5-shim').then(() => import('es6-shim'))` pattern, the fetches are already finished, or at least underway, so there's no worry about sequentializing them, and it just reduces to waiting for any already-under-way fetches, plus sequential execution.

**Comment by ljharb** [comment]
To summarize: I can get the loading I want without the ordering with `Promise.all(paths.map(x => import(x)))`; I can get the ordering I want without the loading with `paths.reduce((p, x) => p.then(() => import(x)), Promise.resolve())`, but to get both in browsers I'd want something like:

``` js
paths.map(x => synchronouslyAddLinkRelPerhaps).reduce((p, x) => p.then(() => import(x)), Promise.resolve())
```

If I add a link rel dynamically, does it kick off the preload? Or does that only work on pageload?

**Comment by domenic** [comment]
Yes, that will work. But if you're doing it dynamically, don't pollute the DOM; just do `fetch(x)` instead. That will warm up the cache equally well.

**Comment by taion** [comment]
@domenic Regarding:

> This sounds like the correct solution to me, if you can't just use static import syntax directly. It is most in line with what this spec is designed for---dynamically kicking off the load of a subgraph. Presumably here shims.js is not actually "shims.js": it's "entry-file-for-a-part-of-the-app-that-depends-on-some-shims-to-function.js". So it will have a lot more than those two imports.

My specific use case is that I've already loaded the code that needs the global polyfills. At run time, I check whether I need the polyfills, though, and only run the rest of my code after loading those polyfills if so. I don't need to load the code that needs the polyfills, because I already have that code.

**Comment by ljharb** [comment]
aha, ok thanks.

While I do think this is something the language could and maybe should solve, the map fetch reduce approach is sufficient for now. Thanks!
